File: sdks/typescript/src/validation.ts
import {
  METRIC_TYPES,
  PRODUCTS,
  Product,
  STATUSES,
} from "./constants";

export class ValidationError extends Error {
  constructor(msg: string) {
    super(msg);
    this.name = "ValidationError";
  }
}

export function validateUsage(
  product: string,
  metricType: string,
  quantity: number,
  status: string,
): void {
  if (!(PRODUCTS as readonly string[]).includes(product)) {
    throw new ValidationError(
      `invalid product '${product}', allowed=${PRODUCTS.join(",")}`,
    );
  }
  const allowed = METRIC_TYPES[product as Product];
  if (!allowed.includes(metricType)) {
    throw new ValidationError(
      `invalid metric_type '${metricType}' for product '${product}', allowed=${allowed.join(",")}`,
    );
  }
  if (!(STATUSES as readonly string[]).includes(status)) {
    throw new ValidationError(
      `invalid status '${status}', allowed=${STATUSES.join(",")}`,
    );
  }
  if (typeof quantity !== "number" || !isFinite(quantity) || quantity < 0) {
    throw new ValidationError(`quantity must be non-negative finite number`);
  }
}

export function safeStringify(obj: unknown): string {
  try {
    return JSON.stringify(obj ?? {}, (_k, v) => {
      if (typeof v === "bigint") return v.toString();
      return v;
    });
  } catch {
    return "{}";
  }
}

File: sdks/python/eka_usage/validation.py
from .constants import PRODUCTS, METRIC_TYPES, STATUSES


class ValidationError(ValueError):
    pass


def validate_usage(product: str, metric_type: str, quantity: float, status: str) -> None:
    if product not in PRODUCTS:
        raise ValidationError(f"invalid product '{product}', allowed={PRODUCTS}")
    allowed = METRIC_TYPES.get(product, ())
    if metric_type not in allowed:
        raise ValidationError(
            f"invalid metric_type '{metric_type}' for product '{product}', allowed={allowed}"
        )
    if status not in STATUSES:
        raise ValidationError(f"invalid status '{status}', allowed={STATUSES}")
    if not isinstance(quantity, (int, float)) or quantity < 0:
        raise ValidationError(f"quantity must be non-negative number, got {quantity!r}")
